src/aktualizr_secondary/secondary_rpc_test.cc

```diff
@@ -0,0 +1,227 @@
+  in_port_t secondary_port = TestUtils::getFreePortAsInt();
+  Uptane::SecondaryInterface::Ptr ip_secondary;
+
+  // trying to connect to a non-running Secondary and create a corresponding instance on Primary
+  ip_secondary = Uptane::IpUptaneSecondary::connectAndCreate("localhost", secondary_port);
+  EXPECT_EQ(ip_secondary, nullptr);
+
+  // create Primary's secondary without connecting to Secondary
+  ip_secondary = std::make_shared<Uptane::IpUptaneSecondary>("localhost", secondary_port, Uptane::EcuSerial("serial"),
+                                                             Uptane::HardwareIdentifier("hwid"),
+                                                             PublicKey("key", KeyType::kED25519));
+
+  Uptane::RawMetaPack meta_pack{"director-root", "director-target", "image_root",
+                                "image_targets", "image_timestamp", "image_snapshot"};
+
+  // expect failures since the secondary is not running
+  EXPECT_EQ(ip_secondary->getManifest(), Json::Value());
+  EXPECT_FALSE(ip_secondary->sendFirmware("firmware"));
+  EXPECT_FALSE(ip_secondary->putMetadata(meta_pack));
+  EXPECT_EQ(ip_secondary->install(""), data::ResultCode::Numeric::kInternalError);
+}
+
+class SecondaryRpcTestNegative : public ::testing::Test {
+ protected:
+  SecondaryRpcTestNegative()
+      : secondary_server_{msg_dispatcher_, "", 0}, secondary_server_thread_{[&]() { secondary_server_.run(); }} {
+    msg_dispatcher_.registerHandler(AKIpUptaneMes_PR_installReq, [](Asn1Message& in_msg, Asn1Message& out_msg) {
+      (void)in_msg;
+      out_msg.present(AKIpUptaneMes_PR_installResp).installResp()->result = AKInstallationResultCode_ok;
+      // out_msg.installResp()->result = AKInstallationResultCode_ok;
+      return MsgDispatcher::HandleStatusCode::kOk;
+    });
+    secondary_server_.wait_until_running();
+  }
+
+  ~SecondaryRpcTestNegative() {
+    secondary_server_.stop();
+    secondary_server_thread_.join();
+  }
+
+  AKIpUptaneMes_PR sendInstallMsg() {
+    // compose and send a valid message
+    Asn1Message::Ptr req(Asn1Message::Empty());
+    req->present(AKIpUptaneMes_PR_installReq);
+
+    // prepare request message
+    auto req_mes = req->installReq();
+    SetString(&req_mes->hash, "target_name");
+    // send request and receive response, a request-response type of RPC
+    std::pair<std::string, uint16_t> secondary_server_addr{"127.0.0.1", secondary_server_.port()};
+    auto resp = Asn1Rpc(req, secondary_server_addr);
+
+    return resp->present();
+  }
+
+ protected:
+  MsgDispatcher msg_dispatcher_;
+  SecondaryTcpServer secondary_server_;
+  std::thread secondary_server_thread_;
+};
+
+// The given test fails because the Secondary TCP server implementation
+// is a single-threaded, synchronous and blocking hence it cannot accept any new connections
+// until the current one is closed. Therefore, if a client/Primary does not close its socket for some reason then
+// Secondary becomes "unavailable"
```

Comment:
Do you expect that we can fix or work around this issue and then restore this test?

Reply:
In general yes. I am not sure if it makes sense to fix it in the given implementation in the light of potential switch to the mqtt based solution which won't have this issue, but the test I guess is valid for both implementations so can be reused.

Reply:
> That's good to hear. I don't think I understand why it would fail now but not before this PR, though.

Because the server is a single-threaded and synchronous so it does not listen/accept a new connection until a current one is closed. This test fails now and will fail for any new changes until the server concurrency model is changed.

Reply:
But was it multithreaded before? I still don't quite understand why it succeeded before this PR.

Reply:
No, it never succeeded before, this is a new test.

Reply:
D'oh. Sorry, I thought it was copied over. Carry on!